pkg/server/endpoints/node/limiter.go

```diff
@@ -0,0 +1,185 @@
+package node
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/time/rate"
+	"google.golang.org/grpc/peer"
+)
+
+const (
+	// Max burst size
+	attestLimit int = 1
+	csrLimit    int = 500
+	jsrLimit    int = 500
+
+	AttestMsg = iota
+	CSRMsg
+	JSRMsg
+)
+
+type Limiter interface {
+	Limit(ctx context.Context, msgType, count int) error
+}
+
+// Newlimiter returns a new node api rate.Limiter
+func NewLimiter(l logrus.FieldLogger) *limiter {
+	return &limiter{
+		attestRate:   1,
+		csrRate:      500,
+		jsrRate:      500,
+		lastNotified: make(map[string]time.Time),
+		limiters:     make(map[int]map[string]*rate.Limiter),
+		log:          l,
+	}
+}
+
+type limiter struct {
+	// Allowed number of messages per second
+	attestRate rate.Limit
+	csrRate    rate.Limit
+	jsrRate    rate.Limit
+
+	lastNotified map[string]time.Time
+	limiters     map[int]map[string]*rate.Limiter
+	log          logrus.FieldLogger
+	mtx          sync.Mutex
+}
+
+// Limit enforces rate limiting policy by blocking until the specified number of messages can
+// be processed. It introspects the context in order to identify the caller. An error will be
+// returned if the context is cancelled, an invalid msgType is specified, or if the number
+// of messages exceeds the burst limit.
+func (l *limiter) Limit(ctx context.Context, msgType, count int) error {
+	callerID, err := l.callerID(ctx)
+	if err != nil {
+		return err
+	}
+
+	rl, err := l.limiterFor(msgType, callerID)
+	if err != nil {
+		return err
+	}
+
+	res := rl.ReserveN(time.Now(), count)
+	if res.Delay() > 0 {
+		l.notify(callerID, msgType)
+	}
+	if !res.OK() {
+		res.Cancel()
+		return errors.New("limiter: burst size exceeded")
+	}
+
+	// Ensure it is possible to complete requests if a deadline is set
+	deadline, ok := ctx.Deadline()
+	if ok && res.Delay() > time.Until(deadline) {
+		res.Cancel()
+		return errors.New("limiter: throttle delay exceeds deadline")
+	}
+
+	timer := time.NewTimer(res.Delay())
+	defer timer.Stop()
+	select {
+	case <-timer.C:
+		return nil
+	case <-ctx.Done():
+		res.Cancel()
+		return ctx.Err()
+	}
+
+}
+
+func (l *limiter) limiterFor(msgType int, callerID string) (*rate.Limiter, error) {
+	limiters := l.limitersFor(msgType)
```

Comment:
since `limitersFor`  is only ever called by `limiterFor`, and holds the lock for the entire function, you may consider calling it while under the lock (and removing the Lock()/Unlock() inside `limitersFor`.

Reply:
I did this out of concern for code safety... instead, opted to leave a code comment on `limitersFor`, and moved to only taking a lock in `limiterFor`